Spring Cloud 2025.1.2 ConfigServerHealthIndicator No default constructor found

**Spring Cloud 2025.1.2 ConfigServerHealthIndicator No default constructor found**
I upgraded my config server application from Spring Cloud 2025.1.1 to 2025.1.2 having ConfigServerHealthIndicator No default constructor found warning in log

**Description**
Recently after Spring Cloud 2025.1.2 release, I upgraded 3 of my microservices Spring Cloud to latest version, for the first 2 services I upgraded from 2025.1.1 to 2025.1.2, in the log side it seems to working fine. Then, after I upgraded my config server from 2025.1.1 to 2025.1.2, it seems to show warning like below.

> 2026-06-18T10:38:32.623+08:00  WARN 6616 --- [config-server] [           main] .s.c.c.p.ConfigurationPropertiesRebinder : Cannot create default instance of org.springframework.cloud.config.server.config.ConfigServerHealthIndicator for reset; skipping property reset
> 
> org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.config.server.config.ConfigServerHealthIndicator]: No default constructor found
> 	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:143)
> 	at org.springframework.cloud.context.properties.ConfigurationPropertiesRebinder.resetBeanToDefaults(ConfigurationPropertiesRebinder.java:192)
> 	at org.springframework.cloud.context.properties.ConfigurationPropertiesRebinder.rebind(ConfigurationPropertiesRebinder.java:167)
> 	at org.springframework.cloud.context.properties.ConfigurationPropertiesRebinder.rebind(ConfigurationPropertiesRebinder.java:111)
> 	at org.springframework.cloud.autoconfigure.ConfigurationPropertiesRebinderAutoConfiguration.afterSingletonsInstantiated(ConfigurationPropertiesRebinderAutoConfiguration.java:70)
> 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:1147)
> 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:994)
> 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:621)
> 	at org.springframework.boot.web.server.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143)
> 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:756)
> 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:445)
> 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
> 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1365)
> 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1354)
> 	at com.quintiq.tiger.configuration.ConfigServerApplication.main(ConfigServerApplication.java:12)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:106)
> 	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:64)
> 	at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:40)
> Caused by: java.lang.NoSuchMethodException: org.springframework.cloud.config.server.config.ConfigServerHealthIndicator.<init>()
> 	at java.base/java.lang.Class.getConstructor0(Class.java:3585)
> 	at java.base/java.lang.Class.getDeclaredConstructor(Class.java:2754)
> 	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:138)
> 	... 21 common frames omitted

In order to remove this warning log, I tried to create class like below, didn't fixed the issue

> import org.springframework.beans.factory.annotation.Qualifier;
> import org.springframework.cloud.config.server.config.ConfigServerHealthIndicator;
> import org.springframework.cloud.config.server.config.ConfigServerProperties;
> import org.springframework.cloud.config.server.environment.EnvironmentRepository;
> import org.springframework.context.annotation.Bean;
> import org.springframework.context.annotation.Configuration;
> 
> // fix log error below
> // org.springframework.beans.BeanInstantiationException:
> // Failed to instantiate [org.springframework.cloud.config.server.config.ConfigServerHealthIndicator]:
> // No default constructor found
> //
> // Caused by: java.lang.NoSuchMethodException: org.springframework.cloud.config.server.config.ConfigServerHealthIndicator.<init>()
> @Configuration
> public class ConfigServerFixConfiguration {
> 
>     @Bean
>     public ConfigServerHealthIndicator configServerHealthIndicator(
>             @Qualifier("compositeEnvironmentRepository") EnvironmentRepository environmentRepository,
>             ConfigServerProperties properties) {
>         return new ConfigServerHealthIndicator(environmentRepository, properties);
>     }
> }

I tried to use 

> spring.cloud.config.server.health.enabled=false

 in my properties file, doesn't fix the issue either

It doesn't break any functionality of my service as we generally didn't not have any method that using ConfigServerHealthIndicator.

I actually have no clue why does it happen in 1 component but not the other 2 that I have also upgraded to 2025.1.2

--- a/spring-cloud-context/src/main/java/org/springframework/cloud/context/properties/ConfigurationPropertiesRebinder.java
+++ b/spring-cloud-context/src/main/java/org/springframework/cloud/context/properties/ConfigurationPropertiesRebinder.java
@@ -17,6 +17,7 @@
 package org.springframework.cloud.context.properties;
 
 import java.beans.PropertyDescriptor;
+import java.lang.reflect.Constructor;
 import java.lang.reflect.Modifier;
 import java.util.Collection;
 import java.util.Collections;
@@ -201,6 +202,18 @@ public class ConfigurationPropertiesRebinder
 	 */
 	private void resetBeanToDefaults(Object bean) {
 		Class<?> targetClass = AopUtils.getTargetClass(bean);
+		if (!hasDefaultConstructor(targetClass)) {
+			// Beans that have no default constructor (for example constructor-bound beans
+			// or beans with required dependencies) cannot be instantiated to obtain their
+			// defaults, so the reset is skipped. The bean is still re-bound from the
+			// Environment afterwards; only reverting removed properties to their defaults
+			// is skipped.
+			if (logger.isDebugEnabled()) {
+				logger.debug("No default constructor for " + targetClass.getName()
+						+ "; skipping property reset before rebinding");
+			}
+			return;
+		}
 		Object freshInstance;
 		try {
 			freshInstance = BeanUtils.instantiateClass(targetClass);
@@ -211,6 +224,20 @@ public class ConfigurationPropertiesRebinder
 			return;
 		}
 		resetProperties(bean, freshInstance, Collections.newSetFromMap(new IdentityHashMap<>()));
+	}
+
+	/**
+	 * Whether the given type declares a no-argument constructor (of any visibility),
+	 * which is what {@link BeanUtils#instantiateClass(Class)} needs to build a defaults
+	 * template.
+	 */
+	private boolean hasDefaultConstructor(Class<?> type) {
+		for (Constructor<?> constructor : type.getDeclaredConstructors()) {
+			if (constructor.getParameterCount() == 0) {
+				return true;
+			}
+		}
+		return false;
 	}
 
 	private void resetProperties(Object bean, Object defaults, Set<Object> visited) {

--- a/spring-cloud-context/src/test/java/org/springframework/cloud/context/properties/ConfigurationPropertiesRebinderNoDefaultConstructorIntegrationTests.java
+++ b/spring-cloud-context/src/test/java/org/springframework/cloud/context/properties/ConfigurationPropertiesRebinderNoDefaultConstructorIntegrationTests.java
@@ -42,7 +42,7 @@ import static org.assertj.core.api.BDDAssertions.then;
 /**
  * Verifies that a {@code @ConfigurationProperties} bean without a default constructor
  * (for example a bean with a required dependency) can be rebound without failing or
- * logging a warning. See spring-cloud-config gh-3252.
+ * logging a warning.
  *
  * @author Ryan Baxter
  */

--- a/spring-cloud-context/src/test/java/org/springframework/cloud/context/properties/ConfigurationPropertiesRebinderNoDefaultConstructorIntegrationTests.java
+++ b/spring-cloud-context/src/test/java/org/springframework/cloud/context/properties/ConfigurationPropertiesRebinderNoDefaultConstructorIntegrationTests.java
@@ -1,0 +1,128 @@
+/*
+ * Copyright 2012-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.context.properties;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.system.CapturedOutput;
+import org.springframework.boot.test.system.OutputCaptureExtension;
+import org.springframework.boot.test.util.TestPropertyValues;
+import org.springframework.cloud.autoconfigure.ConfigurationPropertiesRebinderAutoConfiguration;
+import org.springframework.cloud.autoconfigure.RefreshAutoConfiguration;
+import org.springframework.cloud.context.properties.ConfigurationPropertiesRebinderNoDefaultConstructorIntegrationTests.TestConfiguration;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.test.annotation.DirtiesContext;
+
+import static org.assertj.core.api.BDDAssertions.then;
+
+/**
+ * Verifies that a {@code @ConfigurationProperties} bean without a default constructor
+ * (for example a bean with a required dependency) can be rebound without failing or
+ * logging a warning. See spring-cloud-config gh-3252.
+ *
+ * @author Ryan Baxter
+ */
+@SpringBootTest(classes = TestConfiguration.class, properties = "test.message=Hello")
+@ExtendWith(OutputCaptureExtension.class)
+public class ConfigurationPropertiesRebinderNoDefaultConstructorIntegrationTests {
+
+	@Autowired
+	private TestProperties properties;
+
+	@Autowired
+	private ConfigurationPropertiesRebinder rebinder;
+
+	@Autowired
+	private ConfigurableEnvironment environment;
+
+	@Test
+	@DirtiesContext
+	public void rebindsWithoutWarningWhenNoDefaultConstructor(CapturedOutput output) {
+		then(this.properties.getMessage()).isEqualTo("Hello");
+		// Change the property and rebind
+		TestPropertyValues.of("test.message=World").applyTo(this.environment);
+		this.rebinder.rebind();
+		// Rebinding still applies the new value
+		then(this.properties.getMessage()).isEqualTo("World");
+		// The required dependency is left intact
+		then(this.properties.getDependency()).isEqualTo("dependency");
+		// No warning is logged about being unable to create a default instance
+		then(output).doesNotContain("Cannot create default instance");
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	@EnableConfigurationProperties
+	@Import({ RefreshConfiguration.RebinderConfiguration.class, PropertyPlaceholderAutoConfiguration.class })
+	protected static class TestConfiguration {
+
+		@Bean
+		protected TestProperties testProperties() {
+			return new TestProperties("dependency");
+		}
+
+	}
+
+	// Hack out a protected inner class for testing
+	protected static class RefreshConfiguration extends RefreshAutoConfiguration {
+
+		@Configuration(proxyBeanMethods = false)
+		protected static class RebinderConfiguration extends ConfigurationPropertiesRebinderAutoConfiguration {
+
+			public RebinderConfiguration(ApplicationContext context) {
+				super(context);
+			}
+
+		}
+
+	}
+
+	@ConfigurationProperties("test")
+	protected static class TestProperties {
+
+		private final String dependency;
+
+		private String message;
+
+		protected TestProperties(String dependency) {
+			this.dependency = dependency;
+		}
+
+		public String getDependency() {
+			return this.dependency;
+		}
+
+		public String getMessage() {
+			return this.message;
+		}
+
+		public void setMessage(String message) {
+			this.message = message;
+		}
+
+	}
+
+}

